### PR TITLE
Add field echo (calculator)

### PR DESCRIPTION
**Pull request by atiro**
Adds support for echoing selected fields in the output.

Modelled very much on FieldExtractor, which may actually do what is needed but seems to be for a particular purpose? (as it needs a 'recordId' field)

Feedback welcome!

### CODE REVIEW

File: src/main/java/de/gwdg/metadataqa/api/calculator/EchoField.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+      }
+    }
+    return List.of(new FieldCounterBasedResult<>(getCalculatorName(), resultMap).withNoCompression());
+  }
+
+  @Override
+  public List<String> getHeader() {
+	List<String> headers = new ArrayList<>();
+	for (String fieldName : schema.getEchoFields().keySet()) {
+      headers.add("echo:" + fieldName);
```
Comment (mielvds) [inline src/main/java/de/gwdg/metadataqa/api/calculator/EchoField.java:67]: if the field is echo'ed, do you need the 'echo:' prefix?

**Comment by mielvds** [comment]
Do I understand correctly that this copies a field (like an id) from the original record in the output? That would seriously help me simplify my wrapping code!

File: src/main/java/de/gwdg/metadataqa/api/calculator/EchoField.java (hunk 1 above)
Comment (atiro) [reply src/main/java/de/gwdg/metadataqa/api/calculator/EchoField.java:67]: Possibly not! Was following the convention of other calculators to prefix the field with the calculator name, defer to @pkiraly.

**Comment by atiro** [comment]
Yep that's it @mielvds, likewise for me for creating some visualisations, I wanted to aggregate results by some of the fields, so I've slightly hijacked the calculator concept to pull the values through.

**Comment by pkiraly** [comment]
Maybe we worked somehow parallely. I recently added a new property into the schema called "extractable". See this example (from https://github.com/pkiraly/metadata-qa-api/blob/main/src/test/resources/configuration/schema/meemoo.yaml):

```yaml
format: csv
fields:
  - name: url
    categories: [MANDATORY]
    extractable: true                   <------ this field is extractable, so it will be part of the output
    rules:
      - pattern: ^https?://.*$
        failureScore: -2
      - minCount: 1
        failureScore: -3
```

I guess this has the same effect that your solution, or do I miss something?

**Comment by atiro** [comment]
I saw that @pkiraly but it seemed to enforce a need for a recordId field? wasn't sure what that was for.

**Comment by pkiraly** [comment]
> I saw that @pkiraly but it seemed to enforce a need for a recordId field? wasn't sure what that was for.

@atiro if it is that is an issue to fix. As far as I remember recordId was or is extracted at least if the extraction was turned on, but other data elements could be also extracted above that. I will check it again and if this requirement is still existing,  and I will remove it. Is there any other difference between field extractor and EchoField?

**Comment by pkiraly** [comment]
A suggestion for the process: 
1) I will merge this pull request, in order that you can start using it
2) I will investigate and fix this recordId issue
3) since at that time there will be two almost identical features I will remove one of them

Is that acceptable?

**Comment by pkiraly** [comment]
@atiro One more thing. Github disabled the running of CI for first time contributors. Now I enabled it, and if found a compilation error. Please check it and fix it.

**Comment by mielvds** [comment]
@pkiraly what is the expected behavior of `extractable: true`? Until now, it wasn't really clear to me what `extractable` does, so thanks for that (this should probably go in the README). But what is for instance the output of `measureAsList()` if a field has `extractable: true`?

**Comment by atiro** [comment]
The only other difference @pkiraly was the other PR (#67) to allow extracted/echoed fields to not be calculated as part of Completeness. I can run locally with a patched version so perfectly happy to wait if you want to update FieldExtractor instead.

**Comment by mielvds** [comment]
> @mielvds Hi, yes, this is the intention. If the field has `extractable: true` property and the field extraction is turned on, then the field value will be extracted. The intention of `measureAsList` is to produce a `List<String>`, where the individual elements will be the field vaules (converted to String). If you would like to get the original data tyoe, you should use ``measureAsListOfObjects`which produces a`List`. I am working on refactoring this part, and when I am done with it I'll update the README as well.

Nice!
So just for my understanding,  if I take the following schema

```yaml
format: csv
fields:
  - name: url
    categories: [MANDATORY]
    extractable: true   
```

and apply it to:

```csv
url
https://github.com/
```

by running `measureAsList()` with `completenessMeasurementEnabled`, the calculator would give me (as a `List<String>`object):

```
https://github.com/;1.0
```

Is that correct?

**Comment by pkiraly** [comment]
@mielvds To make is more explicit, if there are two calculators, the field extractor, and the completeness, the the output should be equal to

```Java
List.of("https://github.com/", "1.0");
```

So the list has two items, one is the value of the url, the other (that is "1.0") represents the completeness number.

**Comment by mielvds** [comment]
> > I saw that @pkiraly but it seemed to enforce a need for a recordId field? wasn't sure what that was for.
> 
> @atiro if it is that is an issue to fix. As far as I remember recordId was or is extracted at least if the extraction was turned on, but other data elements could be also extracted above that. I will check it again and if this requirement is still existing, and I will remove it. Is there any other difference between field extractor and EchoField?

I'm also running into this issue. The problem is the hardcoded "recordId" field name in https://github.com/pkiraly/metadata-qa-api/blob/main/src/main/java/de/gwdg/metadataqa/api/calculator/FieldExtractor.java#L25. Extractable fields should be completely configurable and not require the source to contain a field called 'recordId', because that's impossible to enforce.

**Comment by pkiraly** [comment]
@atiro @mielvds Thanks a lot for your comment and this PR! I fixed the FieldExtractor at #69. See the following code:

```java
    Schema schema = new BaseSchema()
      .setFormat(Format.CSV)
      .addField(new JsonBranch("url").setExtractable())
      .addField(new JsonBranch("name"));

    MeasurementConfiguration config = new MeasurementConfiguration()
      .enableFieldExtractor()
      .disableCompletenessMeasurement();

    CalculatorFacade facade = new CalculatorFacade(config)
      .setSchema(schema)
      .setCsvReader(new CsvReader().setHeader(((CsvAwareSchema) schema).getHeader()));

    String fileName = "src/test/resources/csv/meemoo-simple.csv";

    CSVIterator iterator = new CSVIterator(new CSVReaderHeaderAware(new FileReader(fileName)));
    List<List<String>> result = new ArrayList<>();
    while (iterator.hasNext()) {
      String line = CsvReader.toCsv(iterator.next());
      result.add(facade.measureAsList(line));
    }
    assertEquals(2, result.size());
    assertEquals(List.of("https://neurovault.org/images/384958/"), result.get(0));
    assertEquals(List.of("https://neurovault.org/images/93390/"), result.get(1));
```

The meemoo-simple.csv looks like:
```csv
url,name
https://neurovault.org/images/384958/,massivea uditory lexical decision
https://neurovault.org/images/93390/,Language in the aging brain
```

I guess this fix work well. Please git it a test, and let me know if it does not do what you expect.

**Comment by mielvds** [comment]
@pkiraly that's great! This will massively simplify my commandline application; I will share a usable version soon.

**Comment by atiro** [comment]
I'll close this @pkiraly as you have handled it better!

**Pull request closed without merging**